wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginMetaRetriever.java

```diff
@@ -115,65 +117,85 @@ public PluginMetaRetriever(
+      if (idVersionTagDelimiter > -1) {
+        try {
+          repo = new URI(plugin.substring(0, idVersionTagDelimiter));
+        } catch (URISyntaxException e) {
+          throw new InternalInfrastructureException(
+              "Plugin registry URL is incorrect. Problematic plugin entry:" + plugin);
+        }
+      }
+      String[] idVersion = idVersionString.split(":");
       if (idVersion.length != 2 || idVersion[0].isEmpty() || idVersion[1].isEmpty()) {
         throw new InfrastructureException(
             "Plugin format is illegal. Problematic plugin entry:" + plugin);
       }
-      String key = idVersion[0] + ':' + idVersion[1];
-      if (collectedIdVersion.containsKey(key)) {
+      if (collectedFQNs
+          .stream()
+          .anyMatch(p -> p.getId().equals(idVersion[0]) && p.getVersion().equals(idVersion[1]))) {
         throw new InfrastructureException(
-            format("Invalid Che tooling plugins configuration: plugin %s is duplicated", key));
+            format(
+                "Invalid Che tooling plugins configuration: plugin %s is duplicated",
+                idVersion[0] + ":" + idVersion[1])); // even if different repos
       }
-      collectedIdVersion.put(key, Pair.of(idVersion[0], idVersion[1]));
+      collectedFQNs.add(new PluginFQN(repo, idVersion[0], idVersion[1]));
     }
-    return collectedIdVersion.values();
+    return collectedFQNs;
   }
 
-  private Collection<PluginMeta> getMetas(ArrayList<Pair<String, String>> metasIdsVersions)
+  private Collection<PluginMeta> getMetas(List<PluginFQN> pluginFQNs)
       throws InfrastructureException {
     ArrayList<PluginMeta> metas = new ArrayList<>();
-    for (Pair<String, String> metaIdVersion : metasIdsVersions) {
-      metas.add(getMeta(metaIdVersion.first, metaIdVersion.second));
+    for (PluginFQN pluginFqn : pluginFQNs) {
+      metas.add(getMeta(pluginFqn));
     }
 
     return metas;
   }
 
-  private PluginMeta getMeta(String id, String version) throws InfrastructureException {
+  private PluginMeta getMeta(PluginFQN pluginFQN) throws InfrastructureException {
+    final String id = pluginFQN.getId();
+    final String version = pluginFQN.getVersion();
     try {
-      URI metaURI = pluginRegistry.clone().path(id).path(version).path("meta.yaml").build();
+      UriBuilder metaURIBuilder =
+          pluginFQN.getRegistry() == null
+              ? pluginRegistry.clone()
+              : UriBuilder.fromUri(pluginFQN.getRegistry());
 
+      URI metaURI = metaURIBuilder.path(id).path(version).path("meta.yaml").build();
       PluginMeta meta = getBody(metaURI, PluginMeta.class);
       validateMeta(meta, id, version);
       return meta;
-    } catch (IllegalArgumentException | UriBuilderException e) {
+    } catch (IllegalArgumentException | UriBuilderException | MalformedURLException e) {
       throw new InternalInfrastructureException(
           format("Metadata of plugin %s:%s retrieval failed", id, version));
```

Comment:
The message doesn't seem clear enough. I know that it is not your code, but would you be willing to improve the message by adding some info about the particular error we found?